client/deployments/client.go

```diff
@@ -157,6 +191,87 @@ func listArtifact(a artifactData, detailLevel int) {
 	fmt.Println("--------------------------------------------------------------------------------")
 }
 
+func (c *Client) DirectUpload(
+	artifactPath, token, url string,
+	noProgress bool,
+) error {
+	var bar *pb.ProgressBar
+
+	artifact, err := os.Open(artifactPath)
+	if err != nil {
+		return errors.Wrap(err, "Cannot read artifact file")
+	}
+
+	artifactStats, err := artifact.Stat()
+	if err != nil {
+		return errors.Wrap(err, "Cannot read artifact file stats")
+	}
+
+	// create pipe
+	pR, pW := io.Pipe()
+
+	writer := pW
+
+	req, err := http.NewRequest(http.MethodPut, url, pR)
+	if err != nil {
+		return errors.Wrap(err, "Cannot create request")
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.ContentLength = artifactStats.Size()
+
+	reqDump, _ := httputil.DumpRequest(req, false)
+	log.Verbf("sending request: \n%v", string(reqDump))
+
+	if !noProgress {
+		// create progress bar
+		bar = pb.New64(artifactStats.Size()).
+			Set(pb.Bytes, true).
+			SetRefreshRate(time.Millisecond * 100)
+		bar.Start()
+	}
+
+	go func() {
+		defer pW.Close()
+		defer artifact.Close()
+
+		var w io.Writer
+		w = writer
+		if !noProgress {
+			w = bar.NewProxyWriter(writer)
+		}
+
+		if _, err := io.Copy(w, artifact); err != nil {
+			writer.Close()
+			_ = pR.CloseWithError(err)
+			return
+		}
+
+		writer.Close()
+	}()
+
+	rsp, err := c.client.Do(req)
```

Comment:
The headers from the `Link` object needs to be added to request. This is required to make signed URLs work for Azure blob storage.

Reply:
apparently I had no idea how to do this task. sorry for that. Link object returns empty headers, so we have to parse the query parameters (this is also required for the multipart with AWS, so I have been down this path) and that is what I did in addition to the Link.Headers (which take precedence if set). but please check if it is ok.

Reply:
I'm referring to our API. It returns the `headers` attribute if you're using Azure blob storage. I don't think we have to parse the url parameters.

Reply:
yes, I understood that, but have not tested with azure. okey, I removed the query parameters parsing.